chargeio/infrastructure/Requestor.cs

```diff
@@ -86,10 +84,10 @@ public static string Delete(string url, string secretKey = null)
             }
         }
 
-        private static HttpClient PrepareHttpClient()
+        private static HttpClient PrepareHttpClient(string secretKey)
         {
             var client = new HttpClient();
-            var byteArray = Encoding.ASCII.GetBytes(Configuration.SecretKey + ":");
+            var byteArray = Encoding.ASCII.GetBytes(secretKey + ":");

```

Comment:
Since secretKey is nullable (indeed, that is the default), this is a potential null pointer. Secret key at this point must be non-null/non-empty.

Reply:
The parameter secretKey can not be null at this level. ServiceBase checks for this. If the value is some arbitrary string the API should rightly provide an unauthorized access error. This is also true of Configuration.SecretKey as the default value is not a valid merchant secret. In either case null should never make it to the Requestor class.

https://github.com/affinipay/chargeio-net/pull/13/files/46f539c7ac6e39662d409ee6cb459e1807d25890#diff-8680ffe1499af0fe2f749ff63d84c127R11

Reply:
It is potentially null because the public methods on Requestor can be directly invoked without a secretKey specified, which will default them to null. I understand that ServiceBase checks this, but that's easily bypassed. We should be defensive against allowing invalid scenarios such as this, which is easy to do here by throwing a proper exception if secretKey is null here.

Reply:
Given the desire to be defensive against invalid arguments it would be appropriate to throw an ArgumentException in the case that secretKey is a null reference.
https://msdn.microsoft.com/en-us/library/system.argumentexception(v=vs.110).aspx

Reply:
Additionally, given that Requestor is an internal class is the possibility of its public methods being called with bogus parameters legitimate? I mean, we could do that at the maintainer level within the package but that doesn't even make sense. Everything has been routed through ServiceBase to prevent exactly that type of scenario and there are no existing calls to Requestor methods outside of services that derive from ServiceBase.

Reply:
I don't disagree that the Requestor methods secretKey params should have their default values removed now that the Service is defaulting it at construction -- in the previous PR, I noted that only one of ServiceBase or Requestor should be doing that defaulting, but not both.

Reply:
As to whether an internal class should still be guarding against null pointers -- yes, it should, because you can't predict what code paths may be introduced going forward which may give rise to this. It's always best practice to be defensive about the constraints on inputs, regardless of whether the class is public or internal.